tracer/src/Datadog.Tracer.Native/fault_tolerant_method_duplicator.cpp

```diff
@@ -0,0 +1,348 @@
+                mdToken ptOwner;
+                DWORD reserved = 0;
+                WCHAR genericParamName[1024];
+                ULONG pchName;
+                std::vector<mdToken> constraintTypes;
+                mdGenericParam newGenericParam;
+
+                auto enumGenericParamConstraints = trace::Enumerator<mdGenericParamConstraint>(
+                    [&metadataImport, genericParam](HCORENUM* ptr, mdGenericParamConstraint arr[], ULONG max,
+                                                    ULONG* cnt) -> HRESULT {
+                        return metadataImport->EnumGenericParamConstraints(ptr, genericParam, arr, max, cnt);
+                    },
+                    [&metadataImport](HCORENUM ptr) -> void { metadataImport->CloseEnum(ptr); });
+
+                auto genericParamConstraintsIterator = enumGenericParamConstraints.begin();
+                for (; genericParamConstraintsIterator != enumGenericParamConstraints.end();
+                     genericParamConstraintsIterator = ++genericParamConstraintsIterator)
+                {
+                    const auto genericParamConstraint = *genericParamConstraintsIterator;
+
+                    mdGenericParam genericParam;
+                    mdToken constraintType;
+                    hr = metadataImport->GetGenericParamConstraintProps(genericParamConstraint, &genericParam,
+                                                                        &constraintType);
+
+                    if (FAILED(hr))
+                    {
+                        Logger::Warn("    * GetGenericParamConstraintProps has failed. MethodDef: ",
+                                     shared::TokenStr(&methodDef));
+                        shouldSkipToNextMethod = true;
+                        break;
+                    }
+
+                    constraintTypes.push_back(constraintType);
+                }
+
+                if (shouldSkipToNextMethod)
+                {
+                    break;
+                }
+
+                hr = metadataImport->GetGenericParamProps(genericParam, &pulParamSeq, &pdwParamFlags, &ptOwner,
+                                                          &reserved, genericParamName, 1024, &pchName);
+
+                if (FAILED(hr))
+                {
+                    Logger::Warn("    * GetGenericParamProps has failed. MethodDef: ", shared::TokenStr(&methodDef));
+                    shouldSkipToNextMethod = true;
+                    break;
+                }
+
+                if (!constraintTypes.empty())
+                {
+                    std::unique_ptr<mdToken[]> rtkConstraints = std::make_unique<mdToken[]>(constraintTypes.size() + 1);
+                    std::copy(constraintTypes.begin(), constraintTypes.end(), rtkConstraints.get());
+                    rtkConstraints[constraintTypes.size()] = 0;
+                    hr = metadataEmit->DefineGenericParam(instrumentedTargetMethodDef, pulParamSeq, pdwParamFlags,
```

Comment:
Can you call `DefineGenericParam` for both the "Instrumented" method and the "Original" method here to reduce code duplication?

Reply:
Mhm, I will have to call `DefineGenericParam` twice in any case since it has affinity to a particular method, and we have two duplicates that we need to deal with.

As a side note, the Fault-Tolerant duplicating & rewriting codes hold here and there duplicated pieces of code here and there. It's nothing too big but I plan to address that in subsequent PR that will be dedicated for cleanups & minor refactoring.